pkgs/servers/http/nginx/nix-etag-1.x.patch

```diff
@@ -0,0 +1,167 @@
+     r->headers_out.content_length_n = of.size;
+     r->headers_out.last_modified_time = of.mtime;
+ 
+-    if (ngx_http_set_etag(r) != NGX_OK) {
++    if (ngx_http_set_etag_p(r, &path) != NGX_OK) {
+         return NGX_HTTP_INTERNAL_SERVER_ERROR;
+     }
+ 
+diff --git a/src/http/ngx_http_core_module.c b/src/http/ngx_http_core_module.c
+index cb49ef74..3ef39380 100644
+--- a/src/http/ngx_http_core_module.c
++++ b/src/http/ngx_http_core_module.c
+@@ -1577,37 +1577,88 @@ ngx_http_set_exten(ngx_http_request_t *r)
+     return;
+ }
+ 
+-
+ ngx_int_t
+ ngx_http_set_etag(ngx_http_request_t *r)
++{
++    return ngx_http_set_etag_p(r, NULL);
++}
++
++ngx_int_t
++ngx_http_set_etag_p(ngx_http_request_t *r, ngx_str_t *path)
+ {
+     ngx_table_elt_t           *etag;
+     ngx_http_core_loc_conf_t  *clcf;
++    u_char                    *real = NULL, *ptr1, *ptr2;
+ 
+     clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
+ 
+     if (!clcf->etag) {
+         return NGX_OK;
+     }
++    if (r->headers_out.last_modified_time == 1 && path != NULL) {
++        real = ngx_realpath(path->data, NULL);
++
++        if (real == NULL) {
++            return NGX_OK;
++        }
++    }
+ 
+     etag = ngx_list_push(&r->headers_out.headers);
+     if (etag == NULL) {
++        ngx_free(real);
+         return NGX_ERROR;
+     }
+ 
+     etag->hash = 1;
+     ngx_str_set(&etag->key, "ETag");
+ 
+-    etag->value.data = ngx_pnalloc(r->pool, NGX_OFF_T_LEN + NGX_TIME_T_LEN + 3);
+-    if (etag->value.data == NULL) {
+-        etag->hash = 0;
+-        return NGX_ERROR;
+-    }
++    #define NIX_STORE_DIR "/nix/store"
++    #define NIX_STORE_LEN 10
++
++    if (real != NULL
++        && !ngx_strncmp(real, NIX_STORE_DIR, NIX_STORE_LEN)
++        && real[NIX_STORE_LEN] == '/'
++        && real[NIX_STORE_LEN + 1] != '\0')
++    {
++        ptr1 = real + NIX_STORE_LEN;
++        *ptr1 = '"';
++
++        ptr2 = (u_char *) ngx_strchr(ptr1, '-');
+ 
+-    etag->value.len = ngx_sprintf(etag->value.data, "\"%xT-%xO\"",
+-                                  r->headers_out.last_modified_time,
+-                                  r->headers_out.content_length_n)
+-                      - etag->value.data;
++        if (ptr2 == NULL) {
++            ngx_free(real);
++            etag->hash = 0;
++            return NGX_ERROR;
++        }
++
++        *ptr2++ = '"';
++        *ptr2 = '\0';
++
++        etag->value.len = ngx_strlen(ptr1);
++        etag->value.data = ngx_pnalloc(r->pool, etag->value.len);
++
++        if (etag->value.data == NULL) {
++            ngx_free(real);
++            etag->hash = 0;
++            return NGX_ERROR;
++        }
++
++        ngx_memcpy(etag->value.data, ptr1, etag->value.len);
++    } else {
++        etag->value.data = ngx_pnalloc(r->pool, NGX_OFF_T_LEN + NGX_TIME_T_LEN + 3);
++
++        if (etag->value.data == NULL) {
++            etag->hash = 0;
++            if (real != NULL) ngx_free(real);
```

Comment:
In other places you don't do a `NULL` check (and apparently at least `glibc` version of `free` is OK with `free(NULL)`) — why is this place different?

Reply:
I do a NULL check everywhere it could be null, which isn't in the other branch of the if

Reply:
`ngx_free` is just an alias for `free` of the corresponding C library, from http://www.open-std.org/JTC1/SC22/wg14/www/docs/n1124.pdf section 7.20.3.2:

> If ptr is a null pointer, no action occurs.

So we don't need to do explicit checks like this.